runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java

```diff
@@ -739,7 +737,9 @@ private static ParDoPayload getParDoPayload(RunnerApi.PTransform parDoPTransform
 
   public static boolean usesStateOrTimers(AppliedPTransform<?, ?, ?> transform) throws IOException {
     ParDoPayload payload = getParDoPayload(transform);
-    return payload.getStateSpecsCount() > 0 || payload.getTimerFamilySpecsCount() > 0;
+    return payload.getStateSpecsCount() > 0
+        || payload.getTimerFamilySpecsCount() > 0
+        || payload.getRequiresTimeSortedInput();
```

Comment:
That is correct, I'm aware there was no clear consensus, so I tried to minimize the changes needed to fix the underlying issue. I think, that RequiresTimeSortedInput implies "statefulness" (which is the actual purpose of this method), because there is no way to sort data without state (at least in - generic - streaming case). The alternative way seems to be to introduce new `PTransformMatcher` but that would be just for case of "stateless" ordered DoFns, which - although legitimate - are sort of corner case. Vast majority of uses of RequiresTimeSortedInput will match the original condition of `getStateSpecsCount() > 0`. I'm not sure if introducing additional complexity of another PTransformMatcher, incorporating that in all runners that support the annotation is worth it. WDYT?

Reply:
I think modelling how something is executed using the PTransformMatcher makes a lot of sense since not all runners will use state. For example, a runner can sort using their shuffle implementation where the timestamp is the sort key (this is something that Dataflow does for some batch pipelines) and other runners may choose to do this as well.